examples/src/main/java/com/squareup/moshi/recipes/Unwrap.java

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import com.squareup.moshi.recipes.Unwrap.EnvelopJsonAdapter.Enveloped;
+import com.squareup.moshi.recipes.models.Card;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class Unwrap {
+  @Enveloped private static Object unused;
+
+  public static void main(String[] args) throws Exception {
+    String json = ""
+        + "{\"data\":"
+        + "  {\n"
+        + "    \"rank\": \"4\",\n"
+        + "    \"suit\": \"CLUBS\"\n"
+        + "  }"
+        + "}";
+    Annotation annotation = Unwrap.class.getDeclaredField("unused").getAnnotations()[0];
+    Moshi moshi = new Moshi.Builder().add(EnvelopJsonAdapter.FACTORY).build();
+    JsonAdapter<Card> adapter =
+        moshi.adapter(Card.class, Collections.<Annotation>singleton(annotation));
+    Card out = adapter.fromJson(json);
+    System.out.println(out);
+  }
+
+  public static final class EnvelopJsonAdapter extends JsonAdapter<Object> {
+    public static final JsonAdapter.Factory FACTORY = new Factory() {
+      @Override
+      public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+        if (!hasJsonQualifier(annotations, Enveloped.class)) {
+          return null;
+        }
+        Type envelope =
+            Types.newParameterizedTypeWithOwner(EnvelopJsonAdapter.class, Envelope.class,
+                Types.getRawType(type));
+        JsonAdapter<Envelope<?>> delegate =
+            moshi.nextAdapter(this, envelope, Collections.<Annotation>emptySet());
+        return new EnvelopJsonAdapter(delegate);
+      }
+    };
+
+    @Retention(RUNTIME) @JsonQualifier @interface Enveloped {
+    }
+
+    private static final class Envelope<T> {
+      final T data;
+
+      Envelope(T data) {
+        this.data = data;
+      }
+    }
+
+    private final JsonAdapter<Envelope<?>> delegate;
+
+    EnvelopJsonAdapter(JsonAdapter<Envelope<?>> delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override public Object fromJson(JsonReader reader) throws IOException {
+      return delegate.fromJson(reader).data;
+    }
+
+    @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+      delegate.toJson(new Envelope<>(value));
+    }
+  }
+
+  public static boolean hasJsonQualifier(Set<? extends Annotation> annotations,
+      Class<? extends Annotation> jsonQualifier) {
+    if (!jsonQualifier.isAnnotationPresent(JsonQualifier.class)) {
```

Comment:
If this method stays here as a helper, I think this check should be removed, and the check around 107 should just be `annotation instanceof Enveloped`.